### ISSUE TITLE
O3DE 23.05 Integration has an issue in 2.16.5 when AssetBuilder exits abnormally on a baking failure.

### ISSUE THREAD

**Issue by Aleks-Starykh** [PopcornFX/O3DEPopcornFXPlugin#60]
### Environment
The O3DE PKFX Gem v 2.16.5 was successfully imported into an O3DE 23.05 clone project.
(Now, 2.16.5 was registered and integrated into the project flawlessly, unlike 2.16.3-4).
MS VS 2022 17.6.2, cmake called with -G "Visual Studio 17" -DLY_UNITY_BUILD=OFF

### The issue
When the AssetBuilder exits abnormally on a baking failure, the PKFX Gem integration code crashes the AssetBuilder with a `nullptr` exception, because `PopcornFXIntegration::Deactivate()` is called after the `RPISystemInterface`  instance has been deallocated:

```
Prefab Builder: Prefab processing completed.
System: ==================================================================
System: Exception : 0xC0000005 - 'EXCEPTION_ACCESS_VIOLATION' [00007FFE3B14DEF0]
System: <o3de-project>\Gems\PopcornFXPlugin\Code\Source\Integration\PopcornFXIntegration.cpp (156) : PopcornFX::PopcornFXIntegration::Deactivate
System: <o3de-project>\Gems\PopcornFXPlugin\Code\Source\PopcornFXSystemComponent.cpp (162) : PopcornFX::PopcornFXSystemComponent::Deactivate
System: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (242) : AZ::Entity::Deactivate
System: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (114) : AZ::Entity::Reset
System: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (104) : AZ::Entity::~Entity
System: 00007FF71119B114 (AssetBuilder) : AZ::Entity::`scalar deleting destructor'
System: <o3de>\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (870) : AZ::ComponentApplication::Destroy
System: <o3de>\Code\Framework\AzFramework\AzFramework\Application\Application.cpp (273) : AzFramework::Application::Stop
System: <o3de>\Code\Framework\AzToolsFramework\AzToolsFramework\Application\ToolsApplication.cpp (383) : AzToolsFramework::ToolsApplication::Stop
System: <o3de>\Code\Tools\AssetProcessor\AssetBuilder\main.cpp (33) : main
System: D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl (288) : __scrt_common_main_seh
System: 00007FFEECAB7614 (KERNEL32) : BaseThreadInitThunk
System: 00007FFEEDD426F1 (ntdll) : RtlUserThreadStart
E: Exception : 0xC0000005 - 'EXCEPTION_ACCESS_VIOLATION' [00007FFE3B14DEF0]
E: 00007FFEEB4FDDA7 (KERNELBASE) : UnhandledExceptionFilter
E: 00007FFEEDD95570 (ntdll) : memset
E: 00007FFEEDD7C8C6 (ntdll) : _C_specific_handler
E: 00007FFEEDD9247F (ntdll) : _chkstk
E: 00007FFEEDD414F4 (ntdll) : RtlRaiseException
E: 00007FFEEDD90F8E (ntdll) : KiUserExceptionDispatcher
E: <o3de-project>\Gems\PopcornFXPlugin\Code\Source\Integration\PopcornFXIntegration.cpp (156) : PopcornFX::PopcornFXIntegration::Deactivate
E: <o3de-project>\Gems\PopcornFXPlugin\Code\Source\PopcornFXSystemComponent.cpp (162) : PopcornFX::PopcornFXSystemComponent::Deactivate
E: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (242) : AZ::Entity::Deactivate
E: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (114) : AZ::Entity::Reset
E: <o3de>\Code\Framework\AzCore\AzCore\Component\Entity.cpp (104) : AZ::Entity::~Entity
E: 00007FF71119B114 (AssetBuilder) : AZ::Entity::`scalar deleting destructor'
E: <o3de>\Code\Framework\AzCore\AzCore\Component\ComponentApplication.cpp (870) : AZ::ComponentApplication::Destroy
E: <o3de>\Code\Framework\AzFramework\AzFramework\Application\Application.cpp (273) : AzFramework::Application::Stop
E: <o3de>\Code\Framework\AzToolsFramework\AzToolsFramework\Application\ToolsApplication.cpp (383) : AzToolsFramework::ToolsApplication::Stop
E: <o3de>\Code\Tools\AssetProcessor\AssetBuilder\main.cpp (33) : main
E: D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl (288) : __scrt_common_main_seh
E: 00007FFEECAB7614 (KERNEL32) : BaseThreadInitThunk
E: 00007FFEEDD426F1 (ntdll) : RtlUserThreadStart
Exit: Called Terminate() with exit code: 0x1
```
### Temporary hotfix
As a quick hotfix, the following change was applied to [`Code\Source\Integration\PopcornFXIntegration.cpp`](https://github.com/PopcornFX/O3DEPopcornFXPlugin/blob/a4c3e4d08f0cdf2191628830f8ba102fc4a3c9a7/Code/Source/Integration/PopcornFXIntegration.cpp#LL154C23-L154C23)
```
void	PopcornFXIntegration::Deactivate()
{
	//const auto	scene = AZ::RPI::RPISystemInterface::Get()->GetSceneByName(AZ::Name(AzFramework::Scene::MainSceneName));
	if (const auto	systemInterface = AZ::RPI::RPISystemInterface::Get())
	{
		const auto	scene = systemInterface->GetSceneByName(AZ::Name(AzFramework::Scene::MainSceneName));
		if (scene)
		{
			CPopcornFXFeatureProcessor* pkfxFeatureProc = static_cast<CPopcornFXFeatureProcessor*>(scene->GetFeatureProcessor<CPopcornFXFeatureProcessor>());
			if (pkfxFeatureProc)
			{
				pkfxFeatureProc->Init(null, null);
				pkfxFeatureProc->GetRenderManager().Deactivate();
			}
		}
	}
	else
	{
		AZ_ErrorOnce("PopcornFXIntegration", false, "Deactivate() called after RPISystemInterface has been destroyed.");
	}

```
### Expected behaviour
`Deactiveate)()` is being called earlier, when `RPISystemInterface`  instance exists.

**Comment by ValPKFX** [comment]
Good catch thank you !
In fact, this code should not be called since the CPopcornFXFeatureProcessor is not registered in the AssetBuilder as you can see here:
https://github.com/PopcornFX/O3DEPopcornFXPlugin/blob/a4c3e4d08f0cdf2191628830f8ba102fc4a3c9a7/Code/Source/PopcornFXSystemComponent.cpp#L125-L134
I will add the null check on `AZ::RPI::RPISystemInterface::Get()` because it's always better and add an `#if !defined(POPCORNFX_BUILDER)` around all CPopcornFXFeatureProcessor related code in the PopcornFXIntegration.

**Comment by ValPKFX** [comment]
Thank you again for reporting it.
The fix is pushed on the development branch and will be included in the next 2.16 release.
